README.md
# FormMailer

![CircleCI badge](https://img.shields.io/circleci/project/github/Taisiias/formmailer.svg)
[![NPM version][npm-image]][npm-url]

**FormMailer is a simple way to receive emails from contact forms on your static website.**

It runs as a service and emails contents of the forms posted to it.

![Workflow](/img/formmailer-workflow.png)

Suppose you have this HTML in your static page:

```html
<html>
<body>
    <form action="http://myformmailer.mydomain.com/submit" method="post">
        <div>
            <label for="msg">Message:</label>
            <textarea id="msg" name="User Message"></textarea>
        </div>
        <div class="button">
            <button type="submit">Send your message</button>
        </div>
    </form>
</body>
</html>
```

FormMailer service deployed on your domain `http://myformmailer.mydomain.com/` will receive user-submitted POST requests and send emails similar to this one:

```
> Content-Type: text/plain
> From: formmailer@mydomain.com
> To: support@mydomain.com
> Subject: Form submitted on http://mydomain.com/index.html
> Message-ID:
>  <dfc9cdf6-ec60-1a6a-089c-bed4566b05ef@mydomain.com>
> Content-Transfer-Encoding: 7bit
> Date: Mon, 14 Aug 2017 13:29:25 +0000
> MIME-Version: 1.0
>
> Submitted user form was received by FormMailer server, see details below.
>
> Referrer page: http://mydomain.com/index.html
>
> User Message:
>   Hello world!
>
> Submitter IP address: 1.2.3.4
```

FormMailer also:

* supports HTTPS.
* supports plain text and HTML emails.
* uses [Mustache.JS](https://github.com/janl/mustache.js) templates for email body and subject.
* saves all received data (and sent emails) to the local SQLite database.
* understands reCAPTCHA.
* supports `application/json` and `application/x-www-form-urlencoded` content types.

## Running

1. Install [Node.js](https://nodejs.org/en/) (version 8 or higher).

2. Install FormMailer with the command:

    ```bash
    npm install -g formmailer
    ```

3. Create a new `config.json` file and place following defaults inside:

    ```json
    {
        "recipientEmails" : "you@yourdomain.com",
        "fromEmail": "formmailer@localhost"
    }
    ```

     Edit settings specific for your environment (see [Configuration options](#configuration-options)).

4. Start FormMailer:

    ```bash
    formmailer
    ```

    You can specify `config.json` file location with `-c` command line argument: `formmailer -c /path/to/my/config.json`.

Now, change the action field in your HTML form(s) to something like this:
```html
<form method="post" action="http://[domain or ip]:[port]/submit"> ...
```

Here `[domain or ip]` should be your FormMailer server domain (or IP address), and `[port]` should be the port which your FormMailer instance listens to (`httpListenPort` config setting).

## Configuration options

Default configuration file location is `./config.json`. You can provide different location with `-c` command line argument.

Option  | Description | Default
--------|-------------|--------
`recipientEmails` | E-mail recipient address. String or array of strings (for multiple recepients). | Required field.
`fromEmail` | E-mail address that will be provided in `From:` email header. | `"formmailer@localhost"`
`smtpOptions` | [Nodemailer options](https://nodemailer.com/smtp/) object. | `{host: "localhost", port: 25, tls: { rejectUnauthorized: false }}`
`enableHtmlEmail` | Enables sending out HTML versions of emails. | `true`
`enableHttp` | Determines if HTTP should be enabled | `true`
`httpListenIP` | IP address to listen HTTP requests from. | `"0.0.0.0"` (all IP addresses)
`httpListenPort` | Port to listen HTTP requests from. | `80`
`enableHttps` | Determines if HTTPS should be enabled | `true`
`httpsListenIP` | IP address to listen HTTPS requests from. | `"0.0.0.0"` (all IP addresses)
`httpsListenPort` | Port to listen HTTPS requests from. | `443`
`httpsPrivateKeyPath` | Path to HTTPS private key. Specify to enable HTTPS. | `""`
`httpsCertificatePath` | Path to HTTPS certificate. Specify to enable HTTPS. | `""`
`enableRecaptcha` | If false, receiver handler should not check `g-recaptcha-response` field even if site key is provided. | `false`
`reCaptchaSecret` | Site secret reCAPTCHA key. No captcha checks will be performed if this value is not set. | `""`
`reCaptchaSiteKey` | Public reCaptcha site key. | `""`
`formTargets` | See details in [Sending different forms to different recipients](#sending-different-forms-to-different-recipients)| { }
`redirectFieldName` | Name of the HTML input that contains redirect URL address. | `"_redirect"`
`subject` | Email subject field content. Special entry `{{{referrerUrl}}}` will be changed to the address of the webpage from where the form is submitted. | `"Form submitted on {{{referrerUrl}}}"`
`logLevel` | How detailed logging should be (`error`, `warn`, `info`, `verbose`, `debug`, `silly`). | `"info"`
`enableWebInterface` | Enables separate HTTP server that provides the UI for viewing sent email history. | `true`
`webInterfaceIP` | Sent email history HTTP server IP address binding. | `"0.0.0.0"` (all IP addresses)
`webInterfacePort` | Sent email history HTTP server port. | `3002`
`assetsFolder` | Path to the folder containing static assets. | `"./assets"`
`databaseFileName` | Path to the SQLite database file. | `"./formmailer.db"`
`maxHttpRequestSize` | Maximum allowed size of HTTP requests, in bytes. | `1000000`

## Optional features

### Special fields

HTML fields with names that start with the underscore character are ignored by FormMailer:

```html
<input type="text" name="_thisIsIgnored" value="stuff that doesn't have to be sent in email">
```

Also, there are some special fields that can be provided in hidden inputs:

Input name | Meaning
-----------|--------
`_redirect` | URL where to redirect user after the form is successfuly submitted. If `_redirect` field is omitted, user will be redirected to the default `thanks.html` page hosted by FormMailer.
`_formurl` | Value will replace referrer in emails.
`_formname` | Value will show up in email header. Should help to identify which form was submitted if there are several on the page.

### reCAPTCHA

To enable reCAPTCHA checking:

1. Sign up for reCAPTCHA (https://www.google.com/recaptcha/admin), get public site key and secret key.

3. Provide public site key value into `reCaptchaSiteKey` option in your config file.

3. Provide secret key value into `reCaptchaSecret` option in your config file.

There is no need to make any modifications in your site form, reCAPTCHA will be checked on the intermediate page generated by FormMailer.

To disable reCAPTCHA checking, set `enableRecaptcha` to `false`.

If no `reCaptchaSiteKey` is provided in config, you have to integrate reCAPCHA on your site manually, see below.

#### Manual reCAPTCHA

If you are processing reCAPTCHA manually on your site, you need to provide to FormMailer `g-recaptcha-response` field in your POST request (either `x-www-form-urlencoded` or `json`).

Refer to the link below for reCAPTCHA documentation:

https://developers.google.com/recaptcha/docs/display

### Sending different forms to different recipients

First, add a dictionary similar to the one below to the configuration file:

```json
"formTargets": {
    "contact": "contact@mydomain.com",
    "support": ["support@mydomain.com", "team@mydomain.com"],
    "sales": {
        "recipient": "sales@mydomain.com",
        "subject": "Purchase inquiry submitted"
    }
}
```
After that, provide form's action URL in the following format

`http://formmailer.domain.com/submit/<formtarget>` where `<formtarget>` is a key in this dictionary.

```html
<form method="POST" action="http://formmailer.domain.com/submit/sales">...
```

FormMailer will use a corresponding recipient (and optionally a subject) instead of the default one.

### Enabling HTTPS

Acquire SSL certificate (either buy from one of the authorities or [get one for free](https://gethttpsforfree.com/)). You should have two files (usually `.pem` and `.crt`). Put them in some directory.

In the configuration file add private key (`.pem`) path into `httpsPrivateKeyPath` and certificate (`.crt`) path into `httpsCertificatePath`. For example:

```json
{
    ...
    "httpsPrivateKeyPath": "./mycert/mydomain.pem",
    "httpsCertificatePath": "./mycert/mydomain.crt"
}
```

Restart FormMailer and it should automatically run HTTPS server (on port 443), along with HTTP. If you wish to disable HTTP, change `enableHttp` config setting to `false`.

### JSON Resquests

FormMailer accepts JSON requests, distingushing them by `content-type: application/json` HTTP header. FormMailer expects JSON to contain an object. Properties of this object along with their values will be listed in the email. Other behaviour is similar to requests with `content-type: application/x-www-form-urlencoded`, i. e. special fields with names starting with `_` will be ignored and `g-recaptcha-response` field will be used for reCAPTCHA authentication.

If email was sent successfully, JSON response `{ result: "ok" }` will be returned. Or `{ result: "error", description: "error details description" }`, in case of an error.

### Email Templates

FormMailer is using [Mustache.JS](https://github.com/janl/mustache.js) templates for email body and subject. Email can be sent both as plain text and as HTML. Sending out HTML emails can be turned off by setting `enableHtmlEmail` to `false` in the config.

Path to plain text email template is `./assets/plain-text-email-template.mst`.

Path to HTML template is `./assets/html-email-template.html`.

Subject text template can be changed in the `subject` configuration setting.

All templates may include following variables:

Variable | Description
--------|-------------
`formName` | Name of the form (provided by `_formname` special form field).
`incomingIp` | Sender IP address.
`mustacheTemplateData` | User-submitted data (key: `key`, value: `textValue`).
`refererUrl` | Referer URL (from HTTP request headers).

## Viewing Sent Emails History

By default, you can access sent emails history table by the address: `http://localhost:3002/view`. You can disable this by changing `enableWebInterface` setting to `false`.

*NOTE: As there is currently no authentication in the Formmailer web-interface, it is recommended to make `webInterfacePort` (3002 by default) accessible only from local intranet. Or restrict the IP network for web-interface HTTP server by setting `webInterfaceIP` config to something different from `0.0.0.0`.*

## Deploying

**Example for Ubuntu 17.04**

Create a new file `/lib/systemd/system/formmailer.service` and place the following contents inside:

```ini
[Unit]
Description=FormMailer Service
After=network.target

[Service]
Type=simple
User=formmailer
Group=formmailer
ExecStart=/usr/bin/node ./dist/src/index.js -c ./config.json
Restart=on-failure
Environment=NODE_ENV=production
WorkingDirectory=/var/formmailer

[Install]
WantedBy=multi-user.target
```

Change `WorkingDirectory` to the directory where you have cloned FormMailer.

Supposing you have cloned the FormMailer repository to `/var/formmailer`, run:

```bash
$ # create user and group
$ sudo groupadd formmailer && sudo useradd formmailer -g formmailer

$ # go to the formmailer repo directory
$ cd /var/formmailer

$ # change the owner of the formmailer directory to the formmailer user
$ sudo chown formmailer:formmailer . -R

$ # force systemd to load the new service file
$ sudo systemctl daemon-reload

$ # start the service
$ sudo systemctl start formmailer
```

Don't forget to check your firewall settings to allow outside TCP connections to the port specified in `httpListenPort` setting.

*NOTE: FormMailer uses default NodeJS HTTP server. For production environment it is recommended to set up a reverse proxy (Nginx or alternative) that will hide the FormMailer service from the outside world.*

## Alternatives

* [Formspree](https://github.com/formspree/formspree)
* [FormMailerService](https://github.com/abbr/FormMailerService)

## License

[MIT License](LICENSE)

[npm-url]: https://www.npmjs.com/package/formmailer
[npm-image]: https://img.shields.io/npm/v/formmailer.svg


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.2] - 2019-03-31
### Fixed
- Issue #53. Web interface loads indefinitely when accessing email history.

## [1.0.1] - 2018-08-07
### Fixed
- Restored formmailer.js runtime.

## [1.0.0] - 2018-08-07
### Added
- Changed config settings names.
- Default HTTP port is 80.
- Automated testing.

## [0.7.1] - 2018-01-11
### Changed
- Fix: ReCAPTCHA for AJAX requests.
- Fix: Encoding when sending Unicode messages.

## [0.7.0] - 2017-12-18
### Added
- Web interface for viewing email history.
- `enableWebInterface` config setting.
- `webInterfaceIP` config setting.
- `webInterfacePort` config setting.

## [0.6.2] - 2017-11-28
### Changed
- Fix: incorrect package assets folder path.

## [0.6.1] - 2017-11-28
### Added
- Basic sent email history view.

### Changed
- Fix: using package assets if no user assets found.

## [0.6.0] - 2017-11-20
### Added
- Automatic reCaptcha.
- `reCaptchaSiteKey` config setting.
- `disableRecaptcha` config setting.

### Deleted
- `requireReCaptchaResponse` config setting.

## [0.5.0] - 2017-10-23
### Added
- Nodemailer `smtpTransport.SmtpOptions` config setting.
- HTML emails support.

### Deleted
- `smtpPort` config setting.
- `smtpHost` config setting.

## [0.4.0] - 2017-10-6
### Added
- AJAX data request support

## [0.3.3] - 2017-09-19
### Added
- HTTPS support.
- Ability to disable HTTP.

### Changed
- RefererUrl encoding fix.

## [0.3.2] - 2017-09-14
### Changed
- FIX: Start up failure because of incorrect version of node-static dependency.

## [0.3.1] - 2017-09-14
### Added
- Sending different forms to different recipients with `formTargets`.
- Test forms to `index.html`.
- Development config file in `test/config.test.json`.

## [0.3.0] - 2017-09-06
### Added
- Providing form URL in special field `_formurl` (will show up in email instead of referrer).
- Providing form name in special field `_formname` (will show up in emails).
- CONTRIBUTING.

### Changed
- Default URL path for submit handler is `/submit` instead of `/`.

### Deleted
- `httpServerPath` config setting.
- ROADMAP (using GitHub milestones instead).

## [0.2.0] - 2017-09-03
### Added
- Manual reCAPTCHA integration support.
- Installing `formmailer` command in $PATH with `npm install -g formmailer`.
- Configuration options: `reCaptchaSecret`, `requireReCaptchaResponse`.
- CHANGELOG & ROADMAP.

### Changed
- Development environment commands and VSCode debugging integration.
- README fixes.
- Default DB file path changed to `./formmailer.db`

## [0.1.1] - 2017-08-26
### Added
- Example config file.
- CircleCI integration.

### Changed
- Fixed development HTML file and subject template.
- README fixes.

## [0.1.0] - 2017-08-24
### Added
- Basic FormMailer functionality.
- Mustache templating for message body and subject.
- Storing all request information to SQLite database.
- Special `_redirect` form field.
- README.
- First NPM publish.


src/config.ts
import * as deepMerge from "deepmerge";
import * as fs from "fs";
import * as smtpTransport from "nodemailer-smtp-transport";

export interface Config {
    assetsFolder: string;
    httpsCertificatePath: string;
    databaseFileName: string;
    enableHtmlEmail: boolean;
    enableHttp: boolean;
    enableHttps: boolean;
    enableWebInterface: boolean;
    formTargets: { [k: string]: string | string[] | FormTargetData };
    fromEmail: string;
    httpListenIP: string;
    httpsListenIP: string;
    httpListenPort: number;
    httpsListenPort: number;
    logLevel: string;
    httpsPrivateKeyPath: string;
    maxHttpRequestSize: number;
    reCaptchaSecret: string;
    reCaptchaSiteKey: string;
    recipientEmails: string | string[];
    redirectFieldName: string;
    enableRecaptcha: boolean;
    smtpOptions: smtpTransport.SmtpOptions;
    subject: string;
    webInterfaceIP: string;
    webInterfacePort: number;
}

export interface FormTargetData {
    recipient: string | string[];
    subject?: string;
}

const DefaultConfigObject: Config = {
    assetsFolder: "./assets",
    databaseFileName: "./formmailer.db",
    enableHtmlEmail: true,
    enableHttp: true,
    enableHttps: true,
    enableRecaptcha: false,
    enableWebInterface: true,
    formTargets: {},
    fromEmail: "formmailer@localhost",
    httpListenIP: "0.0.0.0",
    httpListenPort: 80,
    httpsCertificatePath: "",
    httpsListenIP: "0.0.0.0",
    httpsListenPort: 443,
    httpsPrivateKeyPath: "",
    logLevel: "info",
    maxHttpRequestSize: 1e6,
    reCaptchaSecret: "",
    reCaptchaSiteKey: "",
    recipientEmails: [],
    redirectFieldName: "_redirect",
    smtpOptions: {
        host: "localhost",
        port: 25,
        tls: { rejectUnauthorized: false },
    },
    subject: "Form submitted on {{{refererUrl}}}",
    webInterfaceIP: "0.0.0.0",
    webInterfacePort: 3002,
};

export function readConfig(path: string): Config {

    let cf: Config;

    if (!fs.existsSync(path)) {
        throw new Error(`Config file was not found.`);
    }

    let fileContent = "";
    try {
        fileContent = fs.readFileSync(path).toString();
    } catch (e) {
        throw new Error(`Config file cannot be read. ${e}`);
    }

    try {
        if (fileContent === "") {
            throw new Error(`Config file is empty`);
        }
        cf = createConfigObject(fileContent);
    } catch (e) {
        throw new Error(`Config file cannot be parsed. ${e}`);
    }
    return cf;
}

export function createConfigObject(fileContent: string): Config {
    let cf: Config;
    const json = JSON.parse(fileContent) as Config;
    /* tslint:disable:no-any */
    const mergedObject: { [k: string]: any } = deepMerge(DefaultConfigObject, json);

    if (!mergedObject.hasOwnProperty("recipientEmails") && !mergedObject.recipientEmails) {
        throw new Error(`Property recipientEmails is missing.`);
    }

    cf = mergedObject as Config;

    return cf;
}
